src/cpp/core/file_lock/LinkBasedFileLock.cpp

```diff
@@ -0,0 +1,318 @@
+   }
+   
+   void clearLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            Error error = lockFilePath.removeIfExists();
+            if (error)
+               LOG_ERROR(error);
+         }
+         registration_.clear();
+      }
+      END_LOCK_MUTEX
+   }
+   
+   ~LockRegistration()
+   {
+      try
+      {
+         clearLocks();
+      }
+      catch (...)
+      {
+      }
+   }
+   
+private:
+   
+   boost::mutex mutex_;
+   std::set<FilePath> registration_;
+   
+};
+
+LockRegistration& lockRegistration()
+{
+   static LockRegistration instance;
+   return instance;
+}
+
+Error writeLockFile(const FilePath& lockFilePath)
+{
+   
+#ifndef _WIN32
+   
+   // generate proxy lockfile
+   FilePath proxyPath = lockFilePath.parent().complete(proxyLockFileName());
+   if (proxyPath.exists())
+      return fileExistsError(ERROR_LOCATION);
+   
+   Error error = core::writeStringToFile(proxyPath, "");
+   if (error)
+      return error;
+   RemoveOnExitScope scope(proxyPath, ERROR_LOCATION);
+   
+   // attempt to link to the desired location -- ignore return value
+   // and just stat our original link after
+   ::link(
+            proxyPath.absolutePathNative().c_str(),
+            lockFilePath.absolutePathNative().c_str());
+   
+   struct stat info;
+   int errc = ::stat(proxyPath.absolutePathNative().c_str(), &info);
+   if (errc)
+      return systemError(errno, ERROR_LOCATION);
+   
+   if (info.st_nlink != 2)
+      return fileExistsError(ERROR_LOCATION);
+   
+   return Success();
+   
+#else
+   
+   return -1;
+   
+#endif
+}
+
+bool isLockFileStale(const FilePath& lockFilePath, double seconds = 30.0)
+{
+   double diff = ::difftime(::time(NULL), lockFilePath.lastWriteTime());
+   return diff >= seconds;
+}
+
+Error beginLocking(const FilePath& lockingPath)
+{
+   // attempt to clean up stale lockfiles
+   if (lockingPath.exists() && isLockFileStale(lockingPath))
+   {
+      Error error = lockingPath.remove();
+      if (error)
+         LOG_ERROR(error);
+   }
+   
+   // write the lock file to take ownership
+   return writeLockFile(lockingPath);
+}
+
+} // end anonymous namespace
+
+struct LinkBasedFileLock::Impl
+{
+   FilePath lockFilePath;
+};
+
+LinkBasedFileLock::LinkBasedFileLock()
+   : pImpl_(new Impl())
+{
+}
+
+LinkBasedFileLock::~LinkBasedFileLock()
+{
+}
+
+FilePath LinkBasedFileLock::lockFilePath() const
+{
+   return pImpl_->lockFilePath;
+}
+
+bool LinkBasedFileLock::isLocked(const FilePath& lockFilePath)
+{
+   if (!lockFilePath.exists())
+      return false;
+   
+   return !isLockFileStale(lockFilePath);
+}
+
+Error LinkBasedFileLock::acquire(const FilePath& lockFilePath)
+{
+   // use filesystem-based mutex to ensure interprocess locking; ie, attempt to
+   // ensure that only one process can attempt to acquire a lock at a time
+   FilePath lockingPath = lockFilePath.parent().complete(".rstudio-locking");
```

Comment:
This just implies that only one process can attempt to acquire a lock in a particular directory at a time -- it's just an extra safe guard for races.

This should (in theory) be unnecessary, but I think the fact that we might either try to claim a stale lockfile, or create a unique lockfile, might cause problems.

We might just want to remove stale lockfiles and do the normal 'create and link' instead of having this re-entrancy guard, but I am not sure if we have any guarantees on the atomicity for removing a file / link...